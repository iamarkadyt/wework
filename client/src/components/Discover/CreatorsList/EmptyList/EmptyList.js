import React from 'react'
import './EmptyList.scss'

const EmptyList = () => (
  <p className="EmptyList__message">Nothing for you right now! Come back later!</p>
)

export default EmptyList
