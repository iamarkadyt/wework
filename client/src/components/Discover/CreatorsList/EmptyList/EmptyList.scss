@import 'variables';

.EmptyList__message {
    background-color: $color-bg-main;
    color: $color-font-secondary;
    padding: .3rem 1rem;
    margin-top: .7rem;
    border-radius: .1rem;
}
