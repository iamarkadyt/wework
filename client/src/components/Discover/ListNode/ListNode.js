import React from 'react'
import { withRouter } from 'react-router-dom'
import { FaUserPlus as IcoAdd } from 'react-icons/fa'
import Field from '../../Field/Field'
import { discoverListNodeType } from '../../../types/index'

const ListNode = withRouter(({
    _id, name, avatar, title, company,
    followAPerson,
    history
}) => (
        <div className="ListNode-container">
            <img className="ListNode-img" src={avatar} alt="" />
            <div className="ListNode-name">
                <Field
                    type="linkButton"
                    label={name}
                    inline
                    style={{ fontSize: '1.15em' }}
                    onClick={() => history.push(`/profile/id/${_id}`)} />
                <p>{title}{company && ` at ${company}`}</p>
            </div>
            <Field
                type="linkButton"
                style={{ fontSize: '1.4rem' }}
                onClick={() => followAPerson(_id)}>
                <IcoAdd />
            </Field>
        </div>
    ))

ListNode.propTypes = {
  node: discoverListNodeType.isRequired
}

export const _UnconnectedListNode = ListNode
export default ListNode
