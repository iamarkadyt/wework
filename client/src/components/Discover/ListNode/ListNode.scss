.ListNode-container {
    margin-top: 1rem;
    display: grid;
    grid-template-columns: 3rem auto 1.4rem;
    grid-column-gap: .5rem;
    align-items: center;
    .ListNode-img {
        width: 3rem;
        border-radius: 3rem;
        height: 3rem;
    }
    .ListNode-name {
        p {
            &:first-child {
                font-size: 1.2rem;
                font-weight: bold;
            }
        }
    }
}
