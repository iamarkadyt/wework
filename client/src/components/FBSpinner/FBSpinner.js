import React from 'react'
import './FBSpinner.scss'

const fBSpinner = () => (
    <div className="lds-facebook">
        <div>
        </div>
        <div>
        </div>
        <div>
        </div>
    </div>
)

export default fBSpinner
