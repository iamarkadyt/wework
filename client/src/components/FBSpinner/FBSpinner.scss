.lds-facebook {
    display: inline-block;
    position: relative;
    width: 3.2rem;
    height: 3.2rem;
    transform: scale(.7);
    margin: 1.5rem 0;
}

.lds-facebook div {
    display: inline-block;
    position: absolute;
    left: .3rem;
    width: .65rem;
    background: rgba($color: #ffffff, $alpha: 0.15);
    animation: lds-facebook 1.2s cubic-bezier(0, 0.5, 0.5, 1) infinite;
}

.lds-facebook div:nth-child(1) {
    left: .3rem;
    animation-delay: -0.24s;
}

.lds-facebook div:nth-child(2) {
    left: 1.3rem;
    animation-delay: -0.12s;
}

.lds-facebook div:nth-child(3) {
    left: 2.25rem;
    animation-delay: 0;
}

@keyframes lds-facebook {
    0% {
        top: .3rem;
        height: 2.55rem;
    }
    50%,
    100% {
        top: .95rem;
        height: 1.3rem;
    }
}