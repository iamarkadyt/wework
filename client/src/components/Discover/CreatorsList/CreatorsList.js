import React, { Fragment } from 'react'
import ListNode from '../ListNode/ListNode'
import { withEither, withAdded } from '../../../hocs/conditionalRendering'
import { compose } from 'recompose'
import FBSpinner from '../../FBSpinner/FBSpinner'
import EmptyList from './EmptyList/EmptyList'
import { arrayOf, func } from 'prop-types'
import { discoverListNodeType } from '../../../types/index'

const CreatorsList = ({
    followAPerson,
    list
}) => (
        <Fragment>
            <h3>Add these people to your feed:</h3>
            <div>
                {list.map(item => {
                    return (
                        <ListNode
                            key={item._id}
                            node={item}
                            followAPerson={followAPerson} />
                    )
                })}
            </div>
        </Fragment>
    )

CreatorsList.propTypes = {
  list: arrayOf(discoverListNodeType).isRequired,
  followAPerson: func.isRequired
}


const isLoadingFn = ({ list }) => !list
const isEmptyFn = ({ list }) => list.length === 0
const withCondRenderings = compose(
  withEither(isLoadingFn, FBSpinner),
  withAdded(isEmptyFn, EmptyList), 
)
const CreatorsListWithCondRenderings = withCondRenderings(CreatorsList)


export const _UnconnectedCreatorsList = CreatorsListWithCondRenderings
export default CreatorsListWithCondRenderings
