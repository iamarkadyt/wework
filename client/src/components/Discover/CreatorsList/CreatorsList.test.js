/**
 * Contract:
 * 
 * What my component renders?
 * - Ver. w/ Snapshot
 *
 * What do I pass to components I render?
 * - div:
 *   - list of ListNodes
 * - each ListNode
 *   - list item,
 *   - followAPerson func
 *
 * What effect do props I receive have?
 * - list turns into an array of ListNodes
 * - if list is undefined/null loading comp is rendered
 * - if list consists of 0 elements, message is shown.
 *
 * What do I do with the state? How do I update it?
 * - N/A
 *
 * Lifecycle hooks effects? Side effects?
 * - N/A
 *
 * Interaction with the component?
 * - N/A
 *
 * Context?
 * - N/A
 *
 * Public ref API?
 * - N/A
 */


import React from 'react'
import { shallow } from 'enzyme'
import { _UnconnectedCreatorsList as CreatorsList } from './CreatorsList'
import ListNode from '../ListNode/ListNode'
import FBSpinner from '../../FBSpinner/FBSpinner'
import EmptyList from './EmptyList/EmptyList'
import cloneDeep from 'lodash.clonedeep'

describe('CreatorsList', () => {
  let props
  let mountedCreatorsList

  const getMockProps = () => {
    return cloneDeep({
      followAPerson: jest.fn(),
      list: [{ 
        _id: 'aaa987',
        name: 'John Doe',
        avatar: '//gravatar.com/12345',
        title: 'Web Developer',
        company: 'Some Company'
      }]
    })
  }
  const creatorsList = () => {
    if(!mountedCreatorsList) {
      mountedCreatorsList = shallow(<CreatorsList {...props} />)
    }
    return mountedCreatorsList
  }

  beforeEach(() => {
    props = getMockProps()
    mountedCreatorsList = undefined
  })

  describe('props passing', () => {
    it('outermost div receives an array of ListNodes as children', () => {
      // dive under cond. rendering hocs
      expect(creatorsList().dive().dive().find('div').find(ListNode).length).toBeGreaterThan(0)
    })

    it('ListNode receives list item and followAPerson func', () => {
      const expectedProps = {
        node: props.list[0],
        followAPerson: expect.any(Function)
      }
      expect(creatorsList().dive().dive().find(ListNode).first().props()).toStrictEqual(expectedProps)
    })
  })

  describe('props effects', () => {
    it('if list isn\'t empty/undefined, it\'s rendered into an array of ListNodes', () => {
      const nodes = 3
      props.list = new Array(nodes).fill(props.list[0]);
      expect(creatorsList().dive().dive().find(ListNode)).toHaveLength(nodes)
    })

    it('if list is empty message is shown', () => {
      props.list = []
      expect(creatorsList().dive().find(EmptyList)).toHaveLength(1)
    })

    it('if list is undefined/null, loading comp is rendered', () => {
      props.list = undefined
      expect(creatorsList().find(FBSpinner)).toHaveLength(1)
    })
  })

  it('must pass a snapshot test', () => {
    expect(creatorsList()).toMatchSnapshot()
  })
})
