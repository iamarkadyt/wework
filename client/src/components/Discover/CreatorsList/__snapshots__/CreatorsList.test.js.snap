// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatorsList must pass a snapshot test 1`] = `
<Component
  followAPerson={[MockFunction]}
  list={
    Array [
      Object {
        "_id": "aaa987",
        "avatar": "//gravatar.com/12345",
        "company": "Some Company",
        "name": "John Doe",
        "title": "Web Developer",
      },
    ]
  }
/>
`;
